helper/impls/helper-impl-110/src/main/java/com/linkedin/avroutil1/compatibility/avro110/codec/AliasAwareSpecificData.java

```diff
@@ -0,0 +1,105 @@
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class AliasAwareSpecificData extends SpecificData {
+
+    //same idea as the one in SpecificData
+    protected final static Map<String, Class<?>> CLASS_CACHE = new ConcurrentHashMap<>();
+    protected final static Class<?> NO_CLASS = (new Object() {}).getClass();
+
+    private final ClassLoader classLoader;//redefined here because older  1.7 versions dont actually have thisd
+
+    public AliasAwareSpecificData() {
+        this.classLoader = null;
+    }
+
+    public AliasAwareSpecificData(ClassLoader classLoader) {
+        super(classLoader);
+        this.classLoader = classLoader;
+    }
+
+    @SuppressWarnings("unchecked")
+    public Object createEnum(String symbol, Schema schema) {
+        @SuppressWarnings("rawtypes")
+        Class c = getClass(schema);
+        if (c == null) {
+            c = lookupByAlias(schema);
+        }
+        if (c == null) {
+            dontPuntToGeneric(schema);
+        }
+        assert c != null; //make IDE happy
+        return Enum.valueOf(c, symbol);
+    }
+
+    @Override
+    public Object createFixed(Object old, Schema schema) {
+        Class<?> c = getClass(schema);
+        if (c == null) {
+            c = lookupByAlias(schema);
+        }
+        if (c == null) {
+            dontPuntToGeneric(schema);
+        }
+        assert c != null; //make IDE happy
+        return c.isInstance(old) ? old : newInstance(c, schema);
+    }
+
+    @Override
+    public Object newRecord(Object old, Schema schema) {
+        Class<?> c = getClass(schema); //this is what vanilla does
+        if (c == null) {
+            c = lookupByAlias(schema);
+        }
+        if (c == null) {
+            dontPuntToGeneric(schema);
+        }
+        assert c != null; //make IDE happy
+        return (c.isInstance(old) ? old : newInstance(c, schema));
+    }
+
+    protected Class<?> lookupByAlias(Schema namedSchema) {
+        Set<String> aliases = namedSchema.getAliases();
+        if (aliases != null) {
+            for (String alias : aliases) {
+                Class<?> byAlias = CLASS_CACHE.computeIfAbsent(alias, fqcn -> {
+                    try {
+                        return ClassLoaderUtil.forName(this.classLoader, alias);
+                    } catch (ClassNotFoundException e) {
+                        return null; //TODO- switch to NO_CLASS
+                    }
+                });
+                if (byAlias != null) {
+                    return byAlias;
+                }
+            }
+        }
+        return null;
+    }
+
+    protected void dontPuntToGeneric(Schema schema) {
+        //vanilla code "punts to generic" at this point and returns a GenericData$Record.
+        //I have never seen this be good for anything except turn this issue into a confusing
+        //ClassCastException downstream, so would rather just throw here
```

Comment:
This makes sense to me, although I wonder if your intent is for this behavior change to be mandatory, or if there is a way to opt-in or out of it? Not an issue either way, but I'm just curious to know how you'll handle it.

Reply:
i throw exceptions on class not found. i never understood what the vanilla behaviour is good for except obfuscating the root cause of failures

Reply:
having said that, this will only kick in if the alias-aware specific decoding is used, which i DO NOT want to make default, and expect to be rare